docs/proposals/zeroconf.md

```diff
@@ -0,0 +1,142 @@
+
+The service will be published to the default ZeroConf domain (`local`) and it's fully-qualifiied domain-name (FQDN) is thus `freddie.local`
+
+> **NOTE** For the purposes of what follows, while distinct, hosts (devices) and services may be considered equivalent.
+
+Then, it's possible to enumerate hosts and services discovered by ZeroConf using:
+
+```bash
+avahi-browse --all
++ wlp6s0 IPv4 freddie                                       _example._tcp        local
+```
+
+## Discovery Process
+
+The protocol implementation is written in Rust and uses [`zeroconf`](https://crates.io/crates/zeroconf). Some of what follows may be specific to these technologies.
+
+> **NOTE** There is a proposal to replace `zeroconf` with [`astro-dnssd`](https://crates.io/crates/astro-dnssd) as this provides cross-platform support. There are some limitations with `astro-dnssd` too that are blocking this switch.
+
+The Akri Agent is deployed to a Kubernetes cluster. Kubernetes clusters commonly run in-cluster DNS services (nowadays [`CoreDNS`](https://kubernetes.io/docs/tasks/administer-cluster/coredns/)). For this reason, the applicability of the Akri ZeroConf protocol is to devices not accessible within the cluster. The benefit of the Akri ZeroConf protocol is to make off-cluster ZeroConf-accessible hosts (devices) and services accessible to Kubernetes cluster resources (e.g. applications).
+
+For ZeroConf discovery to occur, the Agent's Pod must leverage several ZeroConf depdendencies and libraries. These depdendencies not only expand the size of the Akri Agent (~800MB) but they increase the Agent's surface area and increase the possibility of vulnerabilities.
+
+Discovery is a key functionality of ZeroConf and is straightforward to implement. See the [Browsing services](https://crates.io/crates/zeroconf#browsing-services) examples of the `zeroconf` crate.
+
+One wrinkle is that Akri expects discovery to run to completion. Akri periodically reruns discovery for a protocol. The `zeroconf` crate polls networks for hosts and services.
+
+The implementation used by the protocol is to poll for 5 seconds and report back whichever hosts and services were discovered during that window.
+
+## Broker interfacing
+
+Upon detection of ZeroConf hosts and services, the Akri ZeroConf protocol creates "twins" for each service using the provided, sample broker. A more complete rendition of the `freddie` service could be:
+
+```YAML
+{
+    name: "freddie",
+    kind: "_example._tcp",
+    domain: "local",
+    host_name: "freddie.local",
+    address: "192.168.1.100", port: 8888,
+    txt: ...
+}
+```
+
+Each Broker Instance is configured with environment variables corresponding to the above value:
+
+```bash
+AKRI_ZEROCONF_DEVICE_KIND=_example._tcp
+AKRI_ZEROCONF=zeroconf
+AKRI_ZEROCONF_DEVICE_HOST=freddie.local
+AKRI_ZEROCONF_DEVICE_NAME=freddie
+AKRI_ZEROCONF_DEVICE_PORT=8888
+AKRI_ZEROCONF_DEVICE_ADDR=192.168.1.100
+```
+
+Currently the service's TXT records are not provided.
```

Comment:
Are they not provided as a limitation of the zeroconf crate or "not provided" as env vars?

Reply:
My laziness and forgetfulness:

Added: https://github.com/deislabs/akri/pull/162/commits/7f9e77bf7597abe7e323b2b9a35e6d0180a872c1

Set myself a TODO reminder to prototype this.